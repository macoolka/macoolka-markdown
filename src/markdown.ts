/**
 * markdown instance on github for `Document`
 * @desczh
 * `Document`的markdown实例
 * 
 * 基于github
 * @file
 */
import { Document } from 'macoolka-document';
import { head, tail } from 'fp-ts/lib/Array';
import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';

const CRLF = '\n';
const NEWLINE = '\n\n';

const replaceKeys = (a: string) => a.replace(/\|/gi, `\\|`);

/**
 * # h1
 * @since 0.2.0
 */
export const h1: Document['h1'] = title => `# ${title}`;

/**
 * ## h2
 * @since 0.2.0
 */
export const h2: Document['h2'] = title => `## ${title}`;

/**
 * ### h3
 * @since 0.2.0
 */
export const h3: Document['h3'] = title => `### ${title}`;

/**
 * #### h4
 * @since 0.2.0
 */
export const h4: Document['h4'] = title => `#### ${title}`;

/**
 * ##### h5
 * @since 0.2.0
 */
export const h5: Document['h5'] = title => `##### ${title}`;

/**
 * ###### h6
 * @since 0.2.0
 */
export const h6: Document['h6'] = title => `###### ${title}`;
/**
 * **bold**
 * @since 0.2.0
 */
export const bold: Document['bold'] = text => `**${text}**`;
/**
 * *italic*
 * @since 0.2.0
 */
export const italic: Document['italic'] = text => `*${text}*`;
/**
 * ~~del~~
 * @since 0.2.0
 */
export const del: Document['del'] = text => `~~${text}~~`;
/**
 * 'q'
 * @since 0.2.0
 */
export const q: Document['q'] = text => `'${text}'`;
/**
 * **cite**
 * @since 0.2.0
 */
export const cite: Document['cite'] = text => `**${text}**`;
/**
 * **dfn**
 * @since 0.2.0
 */
export const dfn: Document['dfn'] = text => `**${text}**`;
/**
 * **ins**
 * @since 0.2.0
 */
export const ins: Document['ins'] = text => `**${text}**`;
/**
 * **mark**
 * @since 0.2.0
 */
export const mark: Document['mark'] = text => `**${text}**`;
/**
 * **em**
 * @since 0.2.0
 */
export const em: Document['em'] = text => `**${text}**`;
/**
 * 1. ol1
 * 1. ol2"
 * @since 0.2.0
 */
export const ol: Document['ol'] = as => as.map(a => `1. ${a}`).join(CRLF) + NEWLINE;
/**
 * - ul1
 * - ul2"
 *
 *
 * @since 0.2.0
 */
export const ul: Document['ul'] = as => as.map(a => `- ${a}`).join(CRLF) + NEWLINE;
/**
 * article
 *
 *
 * @since 0.2.0
 */
export const article: Document['article'] = text => text + NEWLINE;
/**
 * p
 *
 *
 * @since 0.2.0
 */
export const p: Document['p'] = text => text + NEWLINE;
/**
 * br
 *
 * @since 0.2.0
 */
export const br: Document['br'] = (text) => text + CRLF;
/**
 * [link](/link)
 * @since 0.2.0
 */
export const link: Document['link'] = (text, href) => `[${text}](${href})`;
/**
 * ![img](https://cdn.pixabay.com/photo/2019/08/12/14/23/wedding-4401383_960_720.jpg)
 * @since 0.2.0
 */
export const img: Document['img'] = (text, href) => `![${text}](${href})`;
/**
 * ```ts
 *
 * import a from 'a'
 *
 * ```
 *
 *
 * @since 0.2.0
 */
export const code: Document['code'] = language => a => `\`\`\`${language}${NEWLINE}${a}${NEWLINE}\`\`\`` + NEWLINE;
/**
 * ---
 *
 *
 * @since 0.2.0
 */
export const hr: Document['hr'] = () => '---' + NEWLINE;
/**
 * section
 *
 * @since 0.2.0
 */
export const section: Document['section'] = text => text + NEWLINE;
/**
 * > blockquote 1
 * > blockquote 2"
 *
 *
 * @since 0.2.0
 */
export const blockquote: Document['blockquote'] = as => as.map(a => `> ${a}`).join(CRLF) + NEWLINE;
/**
 * |  head1 | head2  |
 * |  --- | ---  |
 * |  11 | 12  |
 * |  21 | 22  |
 * @since 0.2.0
 */
export const table: Document['table'] = (as: string[][]) => {
    const header = pipe(
        head(as),
        O.map(a =>
            `${CRLF}| ${a.map(replaceKeys).map(v => ` ${v.trim()} `).join('|')} |${CRLF}`
            + `| ${a.map(() => ' --- ').join('|')} |${CRLF}`),
        O.getOrElse(() => '')
    );
    const body = pipe(
        tail(as),
        O.map(
            row =>
                row.map(a =>
                    `| ${a.map(replaceKeys).map(v => ` ${v.trim()} `).join(`|`)} |`).join(CRLF)),
        O.getOrElse(() => '')
    );
    return header + body + CRLF;
};

/**
 * markdown instance on github
 * @desczh
 * 基于github markdown
 */
export const markdown: Document = {
    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    bold,
    italic,
    del,
    q,
    cite,
    dfn,
    ins,
    mark,
    em,
    ol,
    ul,
    article,
    p,
    br,
    link,
    img,
    code,
    hr,
    table,
    section,
    blockquote,

};
