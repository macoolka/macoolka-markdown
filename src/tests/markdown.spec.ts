import {markdown} from '../markdown';

describe('markdown',()=>{
    it('h1',()=>{
        expect(markdown.h1('h1')).toMatchSnapshot();
    })
    it('h2',()=>{
        expect(markdown.h2('h2')).toMatchSnapshot();
    })
    it('h3',()=>{
        expect(markdown.h3('h3')).toMatchSnapshot();
    })
    it('h4',()=>{
        expect(markdown.h4('h4')).toMatchSnapshot();
    })
    it('h5',()=>{
        expect(markdown.h5('h5')).toMatchSnapshot();
    })
    it('h6',()=>{
        expect(markdown.h6('h6')).toMatchSnapshot();
    })
    it('bold',()=>{
        expect(markdown.bold('bold')).toMatchSnapshot();
    })
    it('italic',()=>{
        expect(markdown.italic('italic')).toMatchSnapshot();
    })
    it('del',()=>{
        expect(markdown.del('del')).toMatchSnapshot();
    })
    it('q',()=>{
        expect(markdown.q('q')).toMatchSnapshot();
    })
    it('cite',()=>{
        expect(markdown.cite('cite')).toMatchSnapshot();
    })
    it('dfn',()=>{
        expect(markdown.dfn('dfn')).toMatchSnapshot();
    })
    it('ins',()=>{
        expect(markdown.ins('ins')).toMatchSnapshot();
    })
    it('mark',()=>{
        expect(markdown.mark('mark')).toMatchSnapshot();
    })
    it('em',()=>{
        expect(markdown.em('em')).toMatchSnapshot();
    })
    it('ol',()=>{
        expect(markdown.ol(['ol1','ol2'])).toMatchSnapshot();
    })
    it('ul',()=>{
        expect(markdown.ul(['ul1','ul2'])).toMatchSnapshot();
    })
    it('article',()=>{
        expect(markdown.article('article')).toMatchSnapshot();
    })
    it('p',()=>{
        expect(markdown.p('p')).toMatchSnapshot();
    })
    it('br',()=>{
        expect(markdown.br('br')).toMatchSnapshot();
    })
    it('link',()=>{
        expect(markdown.link('link','/link')).toMatchSnapshot();
    })
    it('img',()=>{
        expect(markdown.img('img','https://www.baidu.com/img/bd_logo1.png')).toMatchSnapshot();
    })
    it('code',()=>{
        expect(markdown.code('ts')(`import a from 'a'`)).toMatchSnapshot();
    })
    it('hr',()=>{
        expect(markdown.hr()).toMatchSnapshot();
    })
    it('section',()=>{
        expect(markdown.section('section')).toMatchSnapshot();
    })
    it('blockquote',()=>{
        expect(markdown.blockquote(['blockquote 1','blockquote 2'])).toMatchSnapshot();
    })
    it('table',()=>{
        expect(markdown.table([['head1','head2'],['11','12'],['21','22']])).toMatchSnapshot();
    })

})