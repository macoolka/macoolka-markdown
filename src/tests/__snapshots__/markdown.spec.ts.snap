// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown article 1`] = `
"article

"
`;

exports[`markdown blockquote 1`] = `
"> blockquote 1
> blockquote 2

"
`;

exports[`markdown bold 1`] = `"**bold**"`;

exports[`markdown br 1`] = `
"br
"
`;

exports[`markdown cite 1`] = `"**cite**"`;

exports[`markdown code 1`] = `
"\`\`\`ts

import a from 'a'

\`\`\`

"
`;

exports[`markdown del 1`] = `"~~del~~"`;

exports[`markdown dfn 1`] = `"**dfn**"`;

exports[`markdown em 1`] = `"**em**"`;

exports[`markdown h1 1`] = `"# h1"`;

exports[`markdown h2 1`] = `"## h2"`;

exports[`markdown h3 1`] = `"### h3"`;

exports[`markdown h4 1`] = `"#### h4"`;

exports[`markdown h5 1`] = `"##### h5"`;

exports[`markdown h6 1`] = `"###### h6"`;

exports[`markdown hr 1`] = `
"---

"
`;

exports[`markdown img 1`] = `"![img](https://www.baidu.com/img/bd_logo1.png)"`;

exports[`markdown ins 1`] = `"**ins**"`;

exports[`markdown italic 1`] = `"*italic*"`;

exports[`markdown link 1`] = `"[link](/link)"`;

exports[`markdown mark 1`] = `"**mark**"`;

exports[`markdown ol 1`] = `
"1. ol1
1. ol2

"
`;

exports[`markdown p 1`] = `
"p

"
`;

exports[`markdown q 1`] = `"'q'"`;

exports[`markdown section 1`] = `
"section

"
`;

exports[`markdown table 1`] = `
"
|  head1 | head2  |
|  --- | ---  |
|  11 | 12  |
|  21 | 22  |
"
`;

exports[`markdown ul 1`] = `
"- ul1
- ul2

"
`;
